web/components/App.jsx

```diff
@@ -0,0 +1,215 @@
+      setFilter(filter.replace(type, ''));
+    } else {
+      setFilter(filter + type);
+    }
+
+    props.getEventsByType(type);
+    setScroll(true);
+  }
+
+  /**
+   * sends a chat message.
+   */
+  async function sendChat() {
+    if (currentText === '') return;
+    setPendingMessage(true);
+
+    try {
+      await messageService.sendMessage(currentText, props.user);
+      setCurrentText('');
+      setPendingMessage(false);
+    } catch (err){
+      console.error(`message failed to send - ${err.message}`);
+    }
+  }
+
+  /**
+   * @param {object} e - event object from file dropped on component
+   */
+  async function uploadAttachment(e) {
+    setFileDraggedOver(false);
+    e.stopPropagation();
+    e.preventDefault();
+
+    try {
+      const data = e.dataTransfer ? e.dataTransfer : e.target;
+      const { files } = data;
+      await attachmentService.postAttachment(files[0], props.selectedChatter);
+    } catch (err){
+      console.error(`failed to upload file - ${err.message}`);
+    }
+  }
+
+  /**
+   * @param {object} e - event object on file dragged over component
+   */
+  function fileDraggedOverHandler(e) {
+    setFileDraggedOver(true);
+    e.stopPropagation();
+    e.preventDefault();
+  }
+
+  return (
+    <div>
+      <FilterHeader filter={filter} changeType={changeFilterType} />
+      <div
+        id="file-drop"
+        onDragOver={fileDraggedOverHandler}
+        className={ fileDraggedOver ? 'file-dragged-over' : 'no-file-dragged-over' }
+        onDragLeave={() => setFileDraggedOver(false)}
+        onDrop={uploadAttachment}
+      >
+        <ul
+          className="slds-chat-list slds-m-bottom--xx-small"
+          id="chat-list"
+          ref={(elem) => {
+            container = elem;
+          }}
+        >
+          <li>
+            <div className="slds-chat-bookend">
+              Start of timeline for&nbsp;<b>Room #{props.roomId}</b>
+            </div>
+          </li>
+          {events.map((event) => {
+            const eventType = event.context && event.context.type;
+            const isAttachment = filter.includes('Attachment') &&
+              eventType === 'Event' && event.context.attachment;
+            const isSelectedFilter = event.context && filter.includes(event.context.type);
+            if (filter === 'All' || isSelectedFilter || isAttachment) {
+              if ((eventType === 'Event' || eventType === 'RoomState') &&
+              !event.context.attachment) {
+                return <EventMessage event={event} key={event.id} />;
+              } else if (eventType === 'User') {
+                return <UserMessage event={event} key={event.id} />;
+              } else if (eventType === 'Event' && event.context.attachment) {
+                return <AttachmentMessage event={event} key={event.id} />;
+              }
+              return <ChatMessage event={event} key={event.id} />;
+            }
+            return <div key={event.id}></div>;
+          })}
```

Comment:
This may only be personal preference, but is it worth extracting this into a separate component or at least into a variable?

Reply:
@MaxGoldschmidt yeah that makes sense. All of this was actually just brought forward from what was there already. I ended up touching a lot of things that weren't related to hooks. But now that you've highlighted it, I will take a look at refactoring this too